### ISSUE TITLE
About EM of squad dataset

### ISSUE THREAD

**Issue by qxp134** [kyutai-labs/ARC-Encoder#4]
I'm really sorry for having so many questions.
Due to time and operational environment constraints, we randomly selected 2000 squad data as the evaluation dataset. We chose ARC8-EncoderM and Mistral 7B decoder as the encoder and decoder combination. The running parameters are: python -m embed_llm.generation.eval_context_comp
--out_file eval.json \
--benchmarks SQUAD \
--run_name ARC8_Encoder_Mistral \
--llm_name mistral_7B \
--llm_number 0 \
--n_icl_exs 5. However, the evaluation result shows an EM of 0.43, which does not match the 0.62 mentioned in the paper. What could be the possible reason for this discrepancy?

<img width="693" height="347" alt="Image" src="https://github.com/user-attachments/assets/705308e6-6ba0-41bb-b2db-ba55c22b31b5" />

**Comment by hippolytepilchen** [comment]
No worries at all, happy to help! 🙂
The discrepancy you observe is expected. The results reported in the paper (e.g., EM ≈ 0.62) correspond to the fine-tuned ARC-Encoder, whereas your setup evaluates the pretrained-only model. An EM around 0.43 for the pretrained ARC8-Encoder with a Mistral 7B decoder on SQuAD is therefore consistent with our own observations.
To give more context, the figure below shows the average performance across several benchmarks (not only SQuAD): the blue curve corresponds to the pretrained encoder, while the green curve shows the gains obtained after fine-tuning. As you can see, fine-tuning leads to a significant improvement, which explains the higher numbers reported in the paper.
So overall, your results look correct given the evaluation setting you described. The code to download the fine-tuning dataset and to fine-tune the model is also available in the repository. 

<img width="461" height="264" alt="Image" src="https://github.com/user-attachments/assets/a2de51c8-1e6c-454e-b4a4-2fbc4232efb7" />

**Comment by qxp134** [comment]
During the fine-tuning training of the encoder, does the large model used by the decoder also need to be involved? This would also require a lot of GPU memory. Compared to fine-tuning the large model of the decoder, what are the advantages?

**Comment by hippolytepilchen** [comment]
Yes, the large decoder model must still be involved during the fine-tuning of the encoder, since it is fed with the compressed tokens produced by the encoder and then used to compute the training loss from its outputs. However, the decoder’s weights are kept frozen and are not updated during this process. While this does imply a significant GPU memory footprint, it can be mitigated by using reduced precision on the decoder.

Compared to fine-tuning the decoder itself, this approach has a key advantage: once the encoder is trained, it can be paired with the exact original decoder (e.g. Mistral 7B) that accepts standard tokens, without any modification to the decoder. As a result, the decoder retains the same performance and behavior as the model released by its authors, while benefiting from the ability to operate on a smaller number of compressed tokens produced by the encoder.

**Comment by qxp134** [comment]
Thank you for your answer to the question.
